### DISCUSSION TITLE
findSymmetry: How to know, and easily adjust the threshold for coordinate errors (e.g. roundoff) in a POSCAR file?

### DISCUSSION THREAD

**Discussion by davidmikolas** [viperleed/viperleed#260] [Q&A] [answered]
I'm just starting to learn ViPErLEED. 

I would like to use findSymmetry on both DFT output (the result of a relaxation) and some atomic coordinates from other relaxations. 

These will sometimes have small deviations from the correct positions for a given symmetry group.  Is there a "close enough" parameter that's used by findSymmetry? Is it easy to adjust?

**Comment by fkraushofer** [comment]
Hi David,

You're probably looking for [SYMMETRY_EPS](https://www.viperleed.org/content/calc/parameters/symmetry_eps.html).

**Comment by davidmikolas** [comment]
@fkraushofer Thanks! I see, we were trying to modify `params.SYMMETRY_EPS` so we could programmatically vary it but that failed. We were thinking that it was a number, but no, it's an object. So now we're wondering if it has a setter method? We've tried dir(params.SYMMETRY_EPS) and suspect there is no setter.

We see that there is a PARAMETERS file, have added a line, but not sure where to put the file or how to confirm if it's being read. `print(params.SYMMETRY_EPS)` shows no change when we change the value in the PARAMETERS file,  currently located in the same folder as the POSCAR file.

**Comment by fkraushofer** [comment]
I'm not sure why you are interacting with the python object at all - that's definitely not the way to set the arguments. Are you calling `viperleed calc`, or are you trying to execute via the ASE interface?

If you're just getting started, the standard, most foolproof way to execute the program is to put all your input files into a folder, then simply execute `viperleed calc` in that folder. If you're completely new to it, I strongly recommend going through the [examples](https://www.viperleed.org/content/calc/examples.html)!

If you're trying to use the symmetry functions by themselves, maybe you could post your script so I can understand what you're trying to do.

EDIT: I'm converting this to a discussion, as it's not really an issue.

**Reply by davidmikolas** [reply]
@fkraushofer  OK. Today we were simply trying to use findSymmetry as a stand-alone utility, as shown in this example https://viperleed.org/content/api.html which turns out to be exactly something that we needed to do today, except that we would like to run it with several values of SYMMETRY_EPS and not just the default. So far, changing the value in the PARAMETERS file has no effect (as seen from `print(params.SYMMETRY_EPS)`) and there seems to be no setter method available to us either. It's late here in Taiwan, my coworker will keep reading and working on it. We'll update here tomorrow. We have plenty of IV-LEED data, and some experience with the relatively simple AquaLEED in the past, but right now we're just getting used to this much bigger and more powerful environment.

**Comment by michele-riva** [comment]
I see fours ways to do what you would like, depending on your use case:
1. Follow a more basic example on how to run a `viperleed calc`, as @fkraushofer pointed out. If all you want to do is "symmetrize" a POSCAR, you need almost nothing in the `PARAMETERS` file: `RUN = 0` and `SYMMETRY_EPS = <your value>`.
2. use the utility for "symmetrizing" a `POSCAR` file. You can find more info [here](https://viperleed.org/content/calc/utilities/poscar_utils.html#enforce-symmetry). **I strongly advise this option or the previous one.**
3. If you really want to use the API, you can create a new `SymmetryEps` value from a floating-point number using the following:

```python
new_eps = param.SYMMETRY_EPS.from_value(<new float value>)
param.SYMMETRY_EPS = new_eps
```
4. Keeping in line with the API example you referred to, you may also use the `Rparams.update(dictionary)` method:

```python
presets = {
    'SYMMETRY_EPS': <new float value>,
    }
params = Rparams()
params.update(presets)
```

**Reply by fkraushofer** [reply]
One comment to option 1.: I'm not sure if this is still the case, but you may need to also generate a dummy `IVBEAMS` file, containing e.g. just `1 0`. But I think for this specific use case, 2. is the most convenient (there is also `viperleed poscar find_symmetry` if you don't want the automatic symmetrization).

**Reply by alexmiame** [reply]
Yes I agree. This is exactly what we made the utilities for :)

**Comment by davidmikolas** [comment]
I am not sure if you both get notifications from this message or not, so I'll do it manually this time (I'm not a frequent Github user) @fkraushofer @amimre @michele-riva 

We had _great success_ with method 3 (API) but not some of the others. I've encouraged my co-worker to join Github soon so he can explain in detail. 

We appreciate all the fast and helpful feedback, and we're looking forward to becoming "power users" :-)

**Reply by alexmiame** [reply]
Happy to hear :)
Let us know if you need any further help!

**Reply by michele-riva** [reply]
One comment concerning the failures you (may have) observed with option 2 in [here](https://github.com/viperleed/viperleed/discussions/260#discussioncomment-10948490): the utility was likely affected by #273, which will be fixed in the upcoming v0.12.2 release.

**Reply by davidmikolas** [reply]
@michele-riva Ah, we'll take a look at that, and also keep and eye out for the new release. I'm curious; is there a mailing list or other way we can get a notification of the release automatically?

**Reply by fkraushofer** [reply]
GitHub lets you receive automatic notifications. On the page of the repository (viperleed/viperleed), simply press the "watch" button (it's an eye symbol), the select "Custom" and choose what you want to be notified about; one of the options is new releases.

**Reply by davidmikolas** [reply]
Ah, I think might have been watching already, but I've switched my option to "all". Does this make me a power-user yet?  <img width="320" alt="Screenshot 2024-11-22 at 2 14 31 PM" src="https://github.com/user-attachments/assets/8b0cd8e6-b402-473f-92a2-661199abd664">

**Reply by fkraushofer** [reply]
Ahaha, I think you'll regret that quite quickly when the emails for every single issue start flooding in; but yeah, that's pretty commited

**Comment by michele-riva** [comment]
@davidmikolas, please let us know if you need more feedback on this. It would be great to hear from your coworker what were the problems you encountered with the "other" alternatives.

If, instead, you feel like this question is resolved, I would ask you to pick one of the comments above as an official answer. This way other people that may have the same question will also see the answer as highlighted.

**Comment by davidmikolas** [comment]
@michele-riva yep yep - I'll or we'll take action within 24 hours, sorry for the delay.

**Comment by KaWeng123** [comment]
Hello, i come here to reply what i do so far, i feel very sorry about the late reply, i am not familiar with Linux and install package, right now i don't even finish the complete installation, very sorry about it. The appendix is what i do so far about installation and the question about read parameter. 
[report.docx](https://github.com/user-attachments/files/17606581/report.docx)

**Reply by fkraushofer** [reply]
Hi, could you please either attach or just paste here  your PARAMETERS file (instead of a screenshot - in yours, all the underscores seem to be missing in the parameter names, but I suspect this is just how the text is rendered). If there are actually no underscores, then this would potentially cause this issue, i.e. replace `SYMMETRY EPS` with `SYMMETRY_EPS` (and the same for all the other parameters). Having the input also makes it easier for us to try and reproduce the issue.

**Reply by KaWeng123** [reply]
Ok, thank you for your quick reply, here is the PARAMETERS file.
[PARAMETERS.txt](https://github.com/user-attachments/files/17607047/PARAMETERS.txt)
It says the original format is not allowed to attach, so i change it to .txt

**Reply by davidmikolas** [reply]
Thanks @KaWeng123 for contributing to our discussion, welcome!  I admire your determination to learn ViPErLEED and start analyzing our data. Also, thanks for joining GitHub!

**Reply by alexmiame** [reply, marked as answer]
Hi!

If you want to set these parameters, using the API, you also have to call parameters.interpret() for it to take effect:
```
from viperleed.calc.files import poscar
from viperleed.calc.files import parameters as vp

slab = poscar.read("POSCAR")  # read slab from POSCAR file
rparams = vp.read("PARAMETERS")  # read parameters from PARAMETERS file
vp.interpret(rparams, slab)  # interpret the parameters file with respect to the slab

print(rparams.SYMMETRY_EPS)  # will print SymmetryEps(0.01)
```

Hope this helps!

**Reply by KaWeng123** [reply]
Thank your for your help, it works.

**Reply by davidmikolas** [reply]
@KaWeng123 if you think you have what you need for now, then (if you see it) you can click "Mark as answer" and that will finish this thread and indicate to others that an answer has been found. You can always ask another question in the future.

If you don't see "Mark as answer" I can do it for you.
